File: capstone-projects/cafe-management-program/java/src/Bill.java
import java.io.*;
import java.util.*;

public class Bill {
    //fields
    private String sID = "BL";
    private String sStaffName;
    Date dDate;
    Time tTime;
    private double dTotal;

    //properties
    public String getsID() {
        return sID;
    }

    public String getsStaffName() {
        return sStaffName;
    }

    public void setsStaffName(String sStaffName) {
        this.sStaffName = sStaffName;
    }

    public Date getdDate() {
        return dDate;
    }

    public void setdDate(Date dDate) {
        this.dDate = dDate;
    }

    public Time gettTime() {
        return tTime;
    }

    public double getdTotal() {
        return dTotal;
    }

    //constructor
    public Bill() {
    }

    public Bill(String id, String StaffName, Date date, Time time, double total) {
        this.sID = id;
        this.sStaffName = StaffName;
        this.dDate = date;
        this.tTime = time;
        this.dTotal = total;
    }

    // methods
    public void Input() {
        if (Cafe.getlBills().size() == 0)
            this.sID += "0";
        else
            this.sID += String.valueOf(FindMaxBillID() + 1);

        System.out.println("Current Bill ID: " + this.sID);
        this.tTime = new Time();
        this.tTime.Input();
    }

    static public int FindMaxBillID() {
        int id = 0;
        for (int i = 0; i < Cafe.getlBills().size(); i++) {
            if (Integer.parseInt(Cafe.getlBills().get(i).getsID().substring(2)) > id)
                id = Integer.parseInt(Cafe.getlBills().get(i).getsID().substring(2));
        }
        return id;
    }

    //Output
    public void Output() {
        System.out.println(Main.padRight("\t" + this.sID, PadRightMax()) +
                Main.padRight(this.sStaffName, PadRightMax()) +
                Main.padRight(Date.ToString(this.dDate), PadRightMax()) +
                Main.padRight(Time.ToString(this.tTime), PadRightMax()) +
                this.dTotal);
    }

    static public void OutputFields() {
        System.out.println(Main.padRight("\n\t[ID]", PadRightMax()) +
                Main.padRight("[STAFF NAME]", PadRightMax()) +
                Main.padRight("[DATE]", PadRightMax()) +
                Main.padRight("[TIME]", PadRightMax()) +
                "[TOTAL]");
    }

    public void CalsTotal(List<Order> loders) {
        this.dTotal = 0;

        for (Order loder : loders) {
            this.dTotal += loder.getdCost();
        }
    }

    //Methods
    static public void SortID() {
        for (int i = 0; i < Cafe.getlBills().size() - 1; i++) {
            for (int j = i + 1; j < Cafe.getlBills().size(); j++) {
                if (Cafe.getlBills().get(i).getsID().compareTo(Cafe.getlBills().get(j).getsID()) > 0) {
                    Bill tmp = Cafe.getlBills().get(i);
                    Cafe.getlBills().set(i, Cafe.getlBills().get(j));
                    Cafe.getlBills().set(j, tmp);
                }
            }
        }
    }

    static public void SortStaffName() {
        for (int i = 0; i < Cafe.getlBills().size() - 1; i++) {
            for (int j = i + 1; j < Cafe.getlBills().size(); j++) {
                if (Cafe.getlBills().get(i).getsStaffName().compareTo(Cafe.getlBills().get(j).getsStaffName()) > 0) {
                    Bill tmp = Cafe.getlBills().get(i);
                    Cafe.getlBills().set(i, Cafe.getlBills().get(j));
                    Cafe.getlBills().set(j, tmp);
                }
            }
        }
    }

    static public void SortDate() {
        for (int i = 0; i < Cafe.getlBills().size() - 1; i++) {
            for (int j = i + 1; j < Cafe.getlBills().size(); j++) {
                if (Date.IsMore(Cafe.getlBills().get(i).getdDate(), Cafe.getlBills().get(j).getdDate())
                        || Date.IsEqual(Cafe.getlBills().get(i).getdDate(), Cafe.getlBills().get(j).getdDate())
                        && Time.IsMore(Cafe.getlBills().get(i).gettTime(), Cafe.getlBills().get(j).gettTime())) {
                    Bill tmp = Cafe.getlBills().get(i);
                    Cafe.getlBills().set(i, Cafe.getlBills().get(j));
                    Cafe.getlBills().set(j, tmp);
                }
            }
        }
    }

    static public void SortTotal() {
        for (int i = 0; i < Cafe.getlBills().size() - 1; i++) {
            for (int j = i + 1; j < Cafe.getlBills().size(); j++) {
                if (Cafe.getlBills().get(i).getdTotal() > (Cafe.getlBills().get(j).getdTotal())) {
                    Bill tmp = Cafe.getlBills().get(i);
                    Cafe.getlBills().set(i, Cafe.getlBills().get(j));
                    Cafe.getlBills().set(j, tmp);
                }
            }
        }
    }

    //Login Methods
    public int FindPadRight() {
        int max = 1;
        while (this.sID.length() > max
                || Date.ToString(this.dDate).length() > max
                || Time.ToString(this.tTime).length() > max
                || String.valueOf(this.dTotal).length() > max) {
            max++;
        }
        return max;
    }

    static public int PadRightMax() {
        int len = Cafe.getlBills().get(0).FindPadRight();
        for (int i = 1; i < Cafe.getlBills().size(); i++) {
            if (Cafe.getlBills().get(i).FindPadRight() > len)
                len = Cafe.getlBills().get(i).FindPadRight();
        }
        return len + 10;
    }

    static public void CalsToTalIncome() {
        for (int i = 0; i < Cafe.getlBills().size(); i++) {
            Cafe.setdTotal(Cafe.getdTotal() + Cafe.getlBills().get(i).getdTotal());
        }
    }

    //FILES
    static public void WriteDataBill() throws IOException {
        String[] wt = new String[Cafe.getlBills().size()];
        for (int i = 0; i < Cafe.getlBills().size(); i++) {
            wt[i] = Cafe.getlBills().get(i).getsID() + ";"
                    + Cafe.getlBills().get(i).getsStaffName() + ";"
                    + Date.ToString(Cafe.getlBills().get(i).getdDate()) + ";"
                    + Time.ToString(Cafe.getlBills().get(i).gettTime()) + ";"
                    + Cafe.getlBills().get(i).getdTotal();
        }

        File file = new File("Bill.txt");
        OutputStream outputStream = new FileOutputStream(file);
        OutputStreamWriter outputStreamWriter = new OutputStreamWriter(outputStream);
        for (String item : wt) {
            outputStreamWriter.write(item);
            // Dùng để xuống hàng
            outputStreamWriter.write("\n");
        }
        outputStreamWriter.flush();
    }

    static public void ReadDataBill() throws IOException {
        File file = new File("Bill.txt");
        InputStream inputStream = new FileInputStream(file);
        InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
        BufferedReader reader = new BufferedReader(inputStreamReader);

        ArrayList<String> a = new ArrayList<>();
        String line;
        while ((line = reader.readLine()) != null) {
            a.add(line);
        }

        for (String s : a) {
            String[] b = s.split(";");
            Bill bl = new Bill(b[0], b[1], Date.ToDate(b[2]), Time.ToTime(b[3]), Double.parseDouble(b[4]));
            Cafe.getlBills().add(bl);
        }

        CalsToTalIncome();
    }
}

File: capstone-projects/cafe-management-program/csharp/Project_Nhom04/Bill.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Project_Nhom04
{
    internal class Bill
    {
		//fields
		private string sID = "BL";
		private string sStaffName;
		Date dDate;
		Time tTime;
		private double dTotal;

		//properties
		public string ID { get { return this.sID; } set { this.sID = value; } }
		public string StaffName { get { return this.sStaffName; } set { this.sStaffName = value; } }
		public Date date { get { return this.dDate; } set { this.dDate = value; } }
		public Time Time { get { return this.tTime; } set { this.tTime = value; } }
		public double Total { get { return this.dTotal; } set { this.dTotal = value; } }

		//constructor
		public Bill() { }

        public Bill(string id, string StaffName, Date date, Time time, double total)
        {
            this.sID = id;
            this.sStaffName = StaffName;
            this.dDate = date;
            this.tTime = time;
            this.dTotal = total;
        }

        public Bill(string id, string StaffName, double total)
        {
			this.sID = id;
			this.sStaffName = StaffName;
			this.dTotal = total;
        }

        public Bill(string id, string StaffName)
        {
            this.sID = id;
            this.sStaffName = StaffName;
        }

        //destructor
        ~Bill() { }

		// methods
		public void Input()
        {
            if (Cafe.lbills.Count() == 0)
                this.sID += (Cafe.lbills.Count()).ToString();
            else
                this.sID += (Convert.ToInt32(FindMaxBillID() + 1)).ToString();

            Console.WriteLine("Current Bill ID: " + this.sID);
			this.tTime = new Time();
            this.tTime.Input();
        }

        static public int FindMaxBillID()
        {
            int id = 0;
            for (int i = 0; i < Cafe.lbills.Count(); i++)
            {
                if (Convert.ToInt32(Cafe.lbills[i].ID.Substring(2)) > id)
                    id = Convert.ToInt32(Cafe.lbills[i].ID.Substring(2));
            }
            return id;
        }

        //Output
        public void Output()
        {
            Console.WriteLine("\t" + this.sID.PadRight(PadRightMax())
                                    + this.StaffName.PadRight(PadRightMax())
                                    + ((string)this.dDate).PadRight(PadRightMax())
                                    + ((string)this.tTime).PadRight(PadRightMax()) 
                                    + this.dTotal);
        }

        static public void OutputFields()
        {
            Console.WriteLine("\n\t" + "[ID]".PadRight(PadRightMax()) 
                                    + "[STAFF NAME]".PadRight(PadRightMax()) 
                                    + "[DATE]".PadRight(PadRightMax()) 
                                    + "[TIME]".PadRight(PadRightMax()) 
                                    + "[TOTAL]");
        }

        public void CalsTotal(List<Order> loders)
		{
			this.dTotal = 0;

            for (int i = 0; i < loders.Count(); i++)
			{
				this.dTotal += loders[i].Cost;

            }
		}

        //Methods
        static public void SortID()
        {
            for (int i = 0; i < Cafe.lbills.Count() - 1; i++)
            {
                for (int j = i + 1; j < Cafe.lbills.Count(); j++)
                {
                    if (String.Compare(Cafe.lbills[i].ID, Cafe.lbills[j].ID) > 0)
                    {
                        Bill tmp = Cafe.lbills[i];
                        Cafe.lbills[i] = Cafe.lbills[j];
                        Cafe.lbills[j] = tmp;
                    }
                }
            }
        }

        static public void SortStaffName()
        {
            for (int i = 0; i < Cafe.lbills.Count() - 1; i++)
            {
                for (int j = i + 1; j < Cafe.lbills.Count(); j++)
                {
                    if (String.Compare(Cafe.lbills[i].StaffName, Cafe.lbills[j].StaffName) > 0)
                    {
                        Bill tmp = Cafe.lbills[i];
                        Cafe.lbills[i] = Cafe.lbills[j];
                        Cafe.lbills[j] = tmp;
                    }
                }
            }
        }

        static public void SortDate()
        {
            for (int i = 0; i < Cafe.lbills.Count() - 1; i++)
            {
                for (int j = i + 1; j < Cafe.lbills.Count(); j++)
                {
                    if (Cafe.lbills[i].date > Cafe.lbills[j].date
                    || (Cafe.lbills[i].date == Cafe.lbills[j].date && Cafe.lbills[i].Time > Cafe.lbills[j].Time))
                    {
                        Bill tmp = Cafe.lbills[i];
                        Cafe.lbills[i] = Cafe.lbills[j];
                        Cafe.lbills[j] = tmp;
                    }
                }
            }
        }

        static public void SortTotal()
        {
            for (int i = 0; i < Cafe.lbills.Count() - 1; i++)
            {
                for (int j = i + 1; j < Cafe.lbills.Count(); j++)
                {
                    if (Cafe.lbills[i].Total > Cafe.lbills[j].Total)
                    {
                        Bill tmp = Cafe.lbills[i];
                        Cafe.lbills[i] = Cafe.lbills[j];
                        Cafe.lbills[j] = tmp;
                    }
                }
            }
        }

        //Login Methods
        public int FindPadRight()
        {
            int max = 1;
            while (this.sID.Length > max
                || this.sStaffName.Length > max
                || ((string)this.dDate).Length > max
                || ((string)this.tTime).Length > max
                || this.dTotal.ToString().Length > max)
            {
                max++;
            }
            return max;
        }

        static public int PadRightMax()
        {
            int len = Cafe.lbills[0].FindPadRight();
            for (int i = 1; i < Cafe.lbills.Count(); i++)
            {
                if (Cafe.lbills[i].FindPadRight() > len)
                    len = Cafe.lbills[i].FindPadRight();
            }
            return len + 10;
        }

        static public void CalsToTalIncome()
        {
            for (int i = 0; i < Cafe.lbills.Count; i++)
            {
                Cafe.total += Cafe.lbills[i].Total;
            }
        }

        //FILES
        static public void WriteDataBill()
        {
            string[] wt = new string[Cafe.lbills.Count()];
            for (int i = 0; i < Cafe.lbills.Count(); i++)
            {
                wt[i] = Cafe.lbills[i].ID
                + ";" + Cafe.lbills[i].StaffName.ToString()
                + ";" + (string)Cafe.lbills[i].date
                + ";" + (string)Cafe.lbills[i].Time
                + ";" + Cafe.lbills[i].Total.ToString();
            }
            CalsToTalIncome();
            File.WriteAllLines("Bill.txt", wt);
        }

        static public void ReadDataBill()
        {
            string[] a = File.ReadAllLines("Bill.txt");

            for (int i = 0; i < a.Length; i++)
            {
                string[] b = a[i].Split(';');
                Bill bl = new Bill(b[0], b[1], (Date)b[2], (Time)b[3], Convert.ToDouble(b[4]));
                Cafe.lbills.Add(bl);
            }

            CalsToTalIncome();
        }
    }	
}
